network/p2p/inspector/validation/control_message_validation_inspector_test.go

```diff
@@ -0,0 +1,450 @@
+		expectedPeerID := peer.ID("peerID987654321")
+		duplicateTopicGraftsReq, err := NewInspectRPCRequest(expectedPeerID, unittest.P2PRPCFixture(unittest.WithGrafts(grafts...)))
+		require.NoError(t, err, "failed to get inspect message request")
+		duplicateTopicPrunesReq, err := NewInspectRPCRequest(expectedPeerID, unittest.P2PRPCFixture(unittest.WithPrunes(prunes...)))
+		require.NoError(t, err, "failed to get inspect message request")
+		duplicateTopicIHavesReq, err := NewInspectRPCRequest(expectedPeerID, unittest.P2PRPCFixture(unittest.WithIHaves(ihaves...)))
+		require.NoError(t, err, "failed to get inspect message request")
+		distributor.On("Distribute", mock.AnythingOfType("*p2p.InvCtrlMsgNotif")).Return(nil).Times(3).Run(func(args mock.Arguments) {
+			notification, ok := args[0].(*p2p.InvCtrlMsgNotif)
+			require.True(t, ok)
+			require.Equal(t, expectedPeerID, notification.PeerID)
+			require.Contains(t, []p2pmsg.ControlMessageType{p2pmsg.CtrlMsgGraft, p2pmsg.CtrlMsgPrune, p2pmsg.CtrlMsgIHave}, notification.MsgType)
+			require.True(t, IsDuplicateTopicErr(notification.Error))
+		})
+		rpcTracker.On("LastHighestIHaveRPCSize").Return(int64(100)).Maybe()
+		require.NoError(t, inspector.processInspectRPCReq(duplicateTopicGraftsReq))
+		require.NoError(t, inspector.processInspectRPCReq(duplicateTopicPrunesReq))
+		require.NoError(t, inspector.processInspectRPCReq(duplicateTopicIHavesReq))
+	})
+
+	t.Run("inspectGraftMessages should disseminate invalid control message notification for invalid graft messages as expected", func(t *testing.T) {
+		inspector, distributor, rpcTracker, _, sporkID := inspectorFixture(t)
+		// create unknown topic
+		unknownTopic, malformedTopic, invalidSporkIDTopic := invalidTopics(sporkID)
+		unknownTopicGraft := unittest.P2PRPCGraftFixture(&unknownTopic)
+		malformedTopicGraft := unittest.P2PRPCGraftFixture(&malformedTopic)
+		invalidSporkIDTopicGraft := unittest.P2PRPCGraftFixture(&invalidSporkIDTopic)
+
+		expectedPeerID := peer.ID("peerID987654321")
+		unknownTopicReq, err := NewInspectRPCRequest(expectedPeerID, unittest.P2PRPCFixture(unittest.WithGrafts(unknownTopicGraft)))
+		require.NoError(t, err, "failed to get inspect message request")
+		malformedTopicReq, err := NewInspectRPCRequest(expectedPeerID, unittest.P2PRPCFixture(unittest.WithGrafts(malformedTopicGraft)))
+		require.NoError(t, err, "failed to get inspect message request")
+		invalidSporkIDTopicReq, err := NewInspectRPCRequest(expectedPeerID, unittest.P2PRPCFixture(unittest.WithGrafts(invalidSporkIDTopicGraft)))
+		require.NoError(t, err, "failed to get inspect message request")
+
+		checkNotif := checkNotifFunc(t, expectedPeerID, p2pmsg.CtrlMsgGraft, channels.IsInvalidTopicErr)
+		distributor.On("Distribute", mock.AnythingOfType("*p2p.InvCtrlMsgNotif")).Return(nil).Times(3).Run(checkNotif)
+		rpcTracker.On("LastHighestIHaveRPCSize").Return(int64(100)).Maybe()
```

Comment:
Please revisit the necessity for this mocking in the sub-tests; for several cases it does not seem required.

Reply:
https://github.com/onflow/flow-go/pull/4695/commits/b3c8840bb53e5e354dc8e61951a2d157c7db730e